supervisor/homeassistant/websocket.py

```diff
@@ -22,41 +26,97 @@ class WSClient:
-        self.ha_version: AwesomeVersion = ha_version
-        self.client: aiohttp.ClientWebSocketResponse = client
-        self.message_id: int = 0
-        self._lock: asyncio.Lock = asyncio.Lock()
+        self.ha_version = ha_version
+        self._client = client
+        self._message_id: int = 0
+        self._loop = loop
+        self._futures: dict[int, asyncio.Future[dict]] = {}
+
+    @property
+    def connected(self) -> bool:
+        """Return if we're currently connected."""
+        return self._client is not None and not self._client.closed
+
+    async def close(self) -> None:
+        """Close down the client."""
+        if not self._client.closed:
+            await self._client.close()
+
+    async def async_send_command(self, message: dict[str, Any]) -> dict | None:
+        """Send a websocket message, and return the response."""
+        self._message_id += 1
+        message["id"] = self._message_id
+        self._futures[message["id"]] = self._loop.create_future()
+        _LOGGER.debug("Sending: %s", message)
+        try:
+            await self._client.send_json(message)
+        except ConnectionError as err:
+            raise HomeAssistantWSConnectionError(err) from err
 
-    async def async_send_command(self, message: dict[str, Any]):
-        """Send a websocket command."""
-        async with self._lock:
-            self.message_id += 1
-            message["id"] = self.message_id
+        try:
+            return await self._futures[message["id"]]
+        finally:
+            self._futures.pop(message["id"])
 
-            _LOGGER.debug("Sending: %s", message)
-            try:
-                await self.client.send_json(message)
-            except ConnectionError:
-                return
+    async def start_listener(self) -> None:
+        """Start listening to the websocket."""
+        if not self.connected:
+            raise HomeAssistantWSConnectionError("Not connected when start listening")
 
-            try:
-                response = await self.client.receive_json()
-            except ConnectionError:
-                return
+        try:
+            while self.connected:
+                await self._receive_json()
+        except HomeAssistantWSConnectionError:
+            pass
 
-            _LOGGER.debug("Received: %s", response)
+        except HomeAssistantWSError as err:
+            _LOGGER.warning(err)
 
-            if response["success"]:
-                return response["result"]
+        finally:
+            await self.close()
 
-            raise HomeAssistantWSError(response)
+    async def _receive_json(self) -> None:
+        """Receive json."""
+        msg = await self._client.receive()
+        _LOGGER.debug("Received: %s", msg)
+
+        if msg.type in (
+            WSMsgType.CLOSE,
+            WSMsgType.CLOSED,
+            WSMsgType.CLOSING,
+            WSMsgType.ERROR,
+        ):
+            raise HomeAssistantWSConnectionError()
```

Comment:
It seems to me we throw graceful connection closes and errors into the same bucket here. Can we differentiate that?

I guess `CLOSE` is the message we get if we are dealing with a regular shutdown.

How about:
```
if msg.type == WSMsgType.CLOSE:
    return False

if msg.type in (
    WSMsgType.CLOSED,
    WSMsgType.CLOSING,
    WSMsgType.ERROR,
):
    raise HomeAssistantWSConnectionError()
```

Reply:
(needs also a `return True` at the end of the function)

Reply:
We can, but that adds logic that is not needed.
In any of these scenarios, we need to start the process of closing down and recreating the client, which should be transparent for the user.

Reply:
Yes, we need to recreate the client in any case.

But I'd still prefer if we differentiate between expected flow (regular close) and error case. I am mainly interested in having a warning in the log or something in the error case. This helps us to understand if the system is behaving as we expect.

Reply:
I don't agree, but whatever. I'll add something